File: src/core/context.ts
import fs from 'fs';
import path from 'path';
import { discoverProject, type MmProject } from './project.js';
import { readGlobalConfig, findChannelByRef } from './config.js';
import { readProjectConfig, type ProjectConfig } from '../db/jsonl.js';
import { normalizeAgentRef } from './agents.js';

export interface ChannelContext {
  project: MmProject;
  channelId: string;
  channelName?: string;
  projectConfig: ProjectConfig | null;
}

function loadProjectConfig(project: MmProject): ProjectConfig | null {
  return readProjectConfig(project.dbPath);
}

function resolveProjectFromPath(rootPath: string): MmProject {
  const dbPath = path.join(rootPath, '.mm', 'mm.db');
  if (!fs.existsSync(dbPath)) {
    throw new Error(`Channel database not found at ${dbPath}`);
  }
  return { root: rootPath, dbPath };
}

export function resolveChannelContext(options?: {
  channel?: string;
  cwd?: string;
}): ChannelContext {
  const channelRef = options?.channel;
  const cwd = options?.cwd ?? process.cwd();
  const globalConfig = readGlobalConfig();

  if (channelRef) {
    const resolved = findChannelByRef(channelRef, globalConfig);
    if (resolved) {
      const project = resolveProjectFromPath(resolved.path);
      const projectConfig = loadProjectConfig(project);
      return {
        project,
        channelId: resolved.id,
        channelName: resolved.name,
        projectConfig,
      };
    }

    try {
      const localProject = discoverProject(cwd);
      const localConfig = loadProjectConfig(localProject);
      if (localConfig?.channel_id &&
        (localConfig.channel_id === channelRef || localConfig.channel_name === channelRef)) {
        return {
          project: localProject,
          channelId: localConfig.channel_id,
          channelName: localConfig.channel_name,
          projectConfig: localConfig,
        };
      }
    } catch {
      // Ignore local lookup errors and fall through
    }

    throw new Error(`Channel not found: ${channelRef}`);
  }

  if (globalConfig?.current_channel) {
    const resolved = findChannelByRef(globalConfig.current_channel, globalConfig);
    if (!resolved) {
      throw new Error(`Current channel not found: ${globalConfig.current_channel}`);
    }
    const project = resolveProjectFromPath(resolved.path);
    const projectConfig = loadProjectConfig(project);
    return {
      project,
      channelId: resolved.id,
      channelName: resolved.name,
      projectConfig,
    };
  }

  const localProject = discoverProject(cwd);
  const localConfig = loadProjectConfig(localProject);
  if (!localConfig?.channel_id) {
    throw new Error('No channel context');
  }

  return {
    project: localProject,
    channelId: localConfig.channel_id,
    channelName: localConfig.channel_name,
    projectConfig: localConfig,
  };
}

export function resolveAgentRef(ref: string, config: ProjectConfig | null): string {
  const normalized = normalizeAgentRef(ref);
  const known = config?.known_agents;
  if (!known) return normalized;

  for (const entry of Object.values(known)) {
    if (entry.name === normalized) {
      return normalized;
    }
    if (entry.global_name === normalized) {
      return entry.name;
    }
    if (entry.nicks && entry.nicks.includes(normalized)) {
      return entry.name;
    }
  }

  return normalized;
}

File: internal/command/context.go
package command

import (
	"database/sql"
	"errors"
	"fmt"
	"os"
	"path/filepath"

	"github.com/adamavenir/mini-msg/internal/core"
	"github.com/adamavenir/mini-msg/internal/db"
	"github.com/spf13/cobra"
)

// CommandContext provides shared command resources.
type CommandContext struct {
	DB            *sql.DB
	Project       core.Project
	JSONMode      bool
	ChannelID     string
	ChannelName   string
	ProjectConfig *db.ProjectConfig
}

// GetContext resolves database and channel context for a command.
func GetContext(cmd *cobra.Command) (*CommandContext, error) {
	projectAlias, _ := cmd.Flags().GetString("project")
	jsonMode, _ := cmd.Flags().GetBool("json")
	channelRef, _ := cmd.Flags().GetString("in")

	if projectAlias != "" {
		mainProject, err := core.DiscoverProject("")
		if err != nil {
			return nil, err
		}
		mainDB, err := db.OpenDatabase(mainProject)
		if err != nil {
			return nil, err
		}
		if err := db.InitSchema(mainDB); err != nil {
			_ = mainDB.Close()
			return nil, err
		}

		linked, err := db.GetLinkedProject(mainDB, projectAlias)
		_ = mainDB.Close()
		if err != nil {
			return nil, err
		}
		if linked == nil {
			return nil, fmt.Errorf("linked project '%s' not found. Use 'mm link' first", projectAlias)
		}
		if _, err := os.Stat(linked.Path); err != nil {
			return nil, fmt.Errorf("linked project '%s' database not found at %s", projectAlias, linked.Path)
		}

		project, err := projectFromDBPath(linked.Path)
		if err != nil {
			return nil, err
		}
		linkedDB, err := db.OpenDatabase(project)
		if err != nil {
			return nil, err
		}
		if err := db.InitSchema(linkedDB); err != nil {
			_ = linkedDB.Close()
			return nil, err
		}

		config, err := db.ReadProjectConfig(project.DBPath)
		if err != nil {
			_ = linkedDB.Close()
			return nil, err
		}

		return &CommandContext{
			DB:            linkedDB,
			Project:       project,
			JSONMode:      jsonMode,
			ProjectConfig: config,
		}, nil
	}

	ctx, err := ResolveChannelContext(channelRef, "")
	if err != nil {
		return nil, err
	}
	conn, err := db.OpenDatabase(ctx.Project)
	if err != nil {
		return nil, err
	}
	if err := db.InitSchema(conn); err != nil {
		_ = conn.Close()
		return nil, err
	}

	return &CommandContext{
		DB:            conn,
		Project:       ctx.Project,
		JSONMode:      jsonMode,
		ChannelID:     ctx.ChannelID,
		ChannelName:   ctx.ChannelName,
		ProjectConfig: ctx.ProjectConfig,
	}, nil
}

// ChannelContext describes resolved channel references.
type ChannelContext struct {
	Project       core.Project
	ChannelID     string
	ChannelName   string
	ProjectConfig *db.ProjectConfig
}

// ResolveChannelContext resolves channel context using global config and local project.
func ResolveChannelContext(channelRef string, cwd string) (*ChannelContext, error) {
	if cwd == "" {
		var err error
		cwd, err = os.Getwd()
		if err != nil {
			return nil, err
		}
	}
	globalConfig, err := core.ReadGlobalConfig()
	if err != nil {
		return nil, err
	}

	if channelRef != "" {
		if id, channel, ok := core.FindChannelByRef(channelRef, globalConfig); ok {
			project, err := projectFromRoot(channel.Path)
			if err != nil {
				return nil, err
			}
			config, err := db.ReadProjectConfig(project.DBPath)
			if err != nil {
				return nil, err
			}
			return &ChannelContext{Project: project, ChannelID: id, ChannelName: channel.Name, ProjectConfig: config}, nil
		}

		if localProject, err := core.DiscoverProject(cwd); err == nil {
			localConfig, err := db.ReadProjectConfig(localProject.DBPath)
			if err != nil {
				return nil, err
			}
			if localConfig != nil && (localConfig.ChannelID == channelRef || localConfig.ChannelName == channelRef) {
				return &ChannelContext{
					Project:       localProject,
					ChannelID:     localConfig.ChannelID,
					ChannelName:   localConfig.ChannelName,
					ProjectConfig: localConfig,
				}, nil
			}
		}

		return nil, fmt.Errorf("channel not found: %s", channelRef)
	}

	localProject, err := core.DiscoverProject(cwd)
	if err != nil {
		return nil, err
	}
	localConfig, err := db.ReadProjectConfig(localProject.DBPath)
	if err != nil {
		return nil, err
	}
	if localConfig == nil || localConfig.ChannelID == "" {
		return nil, errors.New("no channel context: local .mm/ directory has no channel_id")
	}

	return &ChannelContext{
		Project:       localProject,
		ChannelID:     localConfig.ChannelID,
		ChannelName:   localConfig.ChannelName,
		ProjectConfig: localConfig,
	}, nil
}

// ResolveAgentRef resolves agent IDs using known agent aliases.
func ResolveAgentRef(ref string, config *db.ProjectConfig) string {
	normalized := core.NormalizeAgentRef(ref)
	if config == nil || len(config.KnownAgents) == 0 {
		return normalized
	}

	for _, entry := range config.KnownAgents {
		if entry.Name != nil && *entry.Name == normalized {
			return normalized
		}
		if entry.GlobalName != nil && *entry.GlobalName == normalized {
			return derefName(entry.Name, normalized)
		}
		if len(entry.Nicks) > 0 {
			for _, nick := range entry.Nicks {
				if nick == normalized {
					return derefName(entry.Name, normalized)
				}
			}
		}
	}

	return normalized
}

func projectFromRoot(rootPath string) (core.Project, error) {
	dbPath := filepath.Join(rootPath, ".mm", "mm.db")
	if _, err := os.Stat(dbPath); err != nil {
		return core.Project{}, fmt.Errorf("channel database not found at %s", dbPath)
	}
	return core.Project{Root: rootPath, DBPath: dbPath}, nil
}

func projectFromDBPath(dbPath string) (core.Project, error) {
	mmDir := filepath.Dir(dbPath)
	root := filepath.Dir(mmDir)
	if _, err := os.Stat(dbPath); err != nil {
		return core.Project{}, err
	}
	return core.Project{Root: root, DBPath: dbPath}, nil
}

func derefName(value *string, fallback string) string {
	if value == nil {
		return fallback
	}
	return *value
}
